pybamm/experiment/step/base_step.py

```diff
@@ -135,10 +130,42 @@ def __init__(
                 name="Drive Cycle",
             )
             self.period = np.diff(t).min()
+        elif is_python_function:
+            t = pybamm.t - pybamm.InputParameter("start time")
+            self.value = value(t)
+            self.period = _convert_time_to_seconds(period)
         else:
             self.value = value
             self.period = _convert_time_to_seconds(period)
 
+        if (
+            hasattr(self, "calculate_charge_or_discharge")
+            and self.calculate_charge_or_discharge
+        ):
+            direction = self.value_based_charge_or_discharge()
+        self.direction = direction
+
+        # Record all the args for repr and hash
+        self.repr_args = f"{value}, duration={duration}"
+        self.hash_args = f"{value}"
+        if termination:
+            self.repr_args += f", termination={termination}"
+            self.hash_args += f", termination={termination}"
+        if period:
+            self.repr_args += f", period={period}"
+        if temperature:
+            self.repr_args += f", temperature={temperature}"
+            self.hash_args += f", temperature={temperature}"
+        if tags:
+            self.repr_args += f", tags={tags}"
+        if start_time:
+            self.repr_args += f", start_time={start_time}"
+        if description:
+            self.repr_args += f", description={description}"
+        if direction:
+            self.repr_args += f", direction={direction}"
+            self.hash_args += f", direction={direction}"
```

Comment:
Minor but can we break this out into a separate method? Just to start shortening this long code block in the function

Reply:
I agree that this block is a bit awkward, especially because it's placed in the middle of the function where some but not all of the attributes are set. We can put this in its own separate method and make it the final step in initialization, but this is technically breaking (and I'm not certain how breaking it is), and it's unrelated to this issue, so I think this should be its own PR.

Reply:
It would be something like this:
```python
self.repr_args, self.hash_args = self.record_tags(...)
```
Then inside the function there would be no state variables updated inside the function.
```python
def record_tags(...):
        repr_args = f"{value}, duration={duration}"
        hash_args = f"{value}"
        if termination:
            repr_args += f", termination={termination}"
            hash_args += f", termination={termination}"
        if period:
            repr_args += f", period={period}"
        if temperature:
            repr_args += f", temperature={temperature}"
            hash_args += f", temperature={temperature}"
        if tags:
            repr_args += f", tags={tags}"
        if start_time:
            repr_args += f", start_time={start_time}"
        if description:
            repr_args += f", description={description}"
        if direction:
            repr_args += f", direction={direction}"
            hash_args += f", direction={direction}"
        return repr_args, hash_args
```
Not the prettiest, but long functions can be broken up safely without breaking things.

Then if we want to use it as part of initialization, we can do that part of the refactoring later

Reply:
Ok, I just made `record_tags` its own method

Reply:
Thanks